### ISSUE TITLE
Regression in 0.0.74: Self bound not solved for a subclass with a PEP 696 typevar default

### ISSUE THREAD

**Issue by epenet** [astral-sh/ty#4387]
### Summary

Calling a `Self`-returning `classmethod` on a generic subclass that declares a
**default** for its type parameter is rejected in ty 0.0.74. The receiver has the
default applied (`D[float]`), but the type variable in the inherited `Self` bound is
left unsolved (`T@D`), and the two are then compared.

Accepted in 0.0.73, rejected in 0.0.74. pyright accepts it.

### Reproducer (no external deps)

```python
from typing import Self


class Base[T]:
    @classmethod
    def find(cls) -> Self | None:
        return None


class D[T = float](Base[T]):
    pass


D.find()
```

### Actual result (0.0.74)

```
error[invalid-argument-type]: Argument to bound method `Base.find` is incorrect
  --> repro.py:14:1
   |
14 | D.find()
   | ^^^^^^^^ Argument type `D[float]` does not satisfy upper bound `Base[T@D]` of type variable `Self`
```

### Expected result
No diagnostic. `D[float]` is assignable to `Base[float]`; `T@D` should be solved to
the default (`float`) — the same default ty already applied to the receiver.

### Notes

- The default is the trigger. Both of these are accepted on 0.0.74:
  - dropping the default: `class D[T](Base[T])` called as `D.find()` or `D[float].find()`
  - subscripting explicitly: `D[float].find()`
- Not specific to the number of base type parameters — it also reproduces with `class Base[TIT, U, T]` / `class D[T = float](Base[IntTI, float, T])`, which is the shape we hit in real code.

### Impact
Found in [tuya-device-handlers](https://github.com/home-assistant-libs/tuya-device-handlers/pull/494) while bumping ty 0.0.73 → 0.0.74: 45 new false positives, all from one base class whose `find_dpcode` classmethod returns `Self | None`, called on subclasses declaring
defaults (`class DPCodeIntegerWrapper[T = float](DPCodeTypeInformationWrapper[IntegerTypeInformation, float, T])`).

### Workaround
replace `Self` with an explicit method-scoped type variable bound to the base class, which requires a `# noqa: PYI019` since ruff prefers `Self` there.

### Version

ty 0.0.74

**Comment by carljm** [comment]
Good news! This was already fixed in main by astral-sh/ruff#27990, should be released soon.
